File: apps/agent_py/src/lifecoach_agent/contracts/background.py
"""Pydantic mirror of `packages/shared-types/src/background.ts` (ADR 0001).

Contracts for the background / scheduled agent-work subsystem. The TS
package stays the source of truth for the web↔agent wire format; this is
the hand-maintained Python port. `tests/unit/test_background_contracts.py`
ports the same valid/invalid cases so any drift fails CI.

OAuth tokens, raw email bodies, addresses, and full snippets must NEVER
appear in these records — only IDs, counts, status, sanitized error
classes, and short client-safe text.
"""

from __future__ import annotations

import re
from datetime import datetime
from typing import Any, Final, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


def _validate_iso8601(value: str, label: str) -> str:
    """Match Zod's ``z.string().datetime()`` — ISO 8601 with a time
    component (and, like the existing contracts, a timezone via the
    trailing ``Z`` or offset)."""
    try:
        datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError as e:
        raise ValueError(f"invalid ISO 8601 {label}: {value}") from e
    if "T" not in value:
        raise ValueError(f"{label} must include time component")
    return value


# --- workflow kinds -------------------------------------------------------

BACKGROUND_WORKFLOW_KINDS: Final[tuple[str, ...]] = (
    "email_triage_daily",
    "email_urgent_scan",
)

BackgroundWorkflowKind = Literal["email_triage_daily", "email_urgent_scan"]


# --- schedule -------------------------------------------------------------

PERMITTED_ACTION_MODES: Final[tuple[str, ...]] = (
    "never",
    "after_confirmation",
    "auto_if_rule_matches",
)

PermittedActionMode = Literal["never", "after_confirmation", "auto_if_rule_matches"]

LOOKBACK_WINDOWS: Final[tuple[str, ...]] = ("12h", "1d", "3d")

LookbackWindow = Literal["12h", "1d", "3d"]

SCHEDULE_LAST_STATUSES: Final[tuple[str, ...]] = ("ok", "skipped", "failed")

ScheduleLastStatus = Literal["ok", "skipped", "failed"]

_LOCAL_TIME_RE = re.compile(r"^([01]\d|2[0-3]):[0-5]\d$")


class ScheduleCadence(BaseModel):
    """Daily cadence policy. ``localTime`` is ``HH:MM`` 24h in the
    schedule's IANA ``timezone``; ``weekdays`` (0=Sunday … 6=Saturday)
    narrows to specific days — omit for every day."""

    model_config = ConfigDict(extra="forbid")

    type: Literal["daily"]
    localTime: str  # noqa: N815 — wire camelCase preserved for parity
    weekdays: list[int] | None = None

    @field_validator("localTime")
    @classmethod
    def local_time_hhmm(cls, v: str) -> str:
        if not _LOCAL_TIME_RE.match(v):
            raise ValueError(f"localTime must be HH:MM 24h: {v}")
        return v

    @field_validator("weekdays")
    @classmethod
    def weekdays_in_range(cls, v: list[int] | None) -> list[int] | None:
        if v is None:
            return v
        for day in v:
            if day < 0 or day > 6:
                raise ValueError(f"weekday out of range 0..6: {day}")
        return v


class PermittedActions(BaseModel):
    model_config = ConfigDict(extra="forbid")

    archiveNoise: PermittedActionMode  # noqa: N815
    createTasks: PermittedActionMode  # noqa: N815
    createCalendarEvents: PermittedActionMode  # noqa: N815


class NotifyPreferences(BaseModel):
    model_config = ConfigDict(extra="forbid")

    inApp: bool  # noqa: N815
    email: bool
    chatSummaryOnNextOpen: bool  # noqa: N815


class BackgroundSchedule(BaseModel):
    """A user-configured automation. Application state, not model state."""

    model_config = ConfigDict(extra="forbid")

    id: str = Field(min_length=1)
    uid: str = Field(min_length=1)
    kind: BackgroundWorkflowKind
    enabled: bool
    timezone: str = Field(min_length=1)
    cadence: ScheduleCadence
    lookbackWindow: LookbackWindow  # noqa: N815
    consentVersion: str = Field(min_length=1)  # noqa: N815
    permittedActions: PermittedActions  # noqa: N815
    notify: NotifyPreferences
    nextRunAt: str  # noqa: N815
    lastRunAt: str | None = None  # noqa: N815
    lastStatus: ScheduleLastStatus | None = None  # noqa: N815
    createdAt: str  # noqa: N815
    updatedAt: str  # noqa: N815

    @field_validator("nextRunAt", "createdAt", "updatedAt")
    @classmethod
    def required_timestamps(cls, v: str) -> str:
        return _validate_iso8601(v, "timestamp")

    @field_validator("lastRunAt")
    @classmethod
    def optional_timestamp(cls, v: str | None) -> str | None:
        return v if v is None else _validate_iso8601(v, "lastRunAt")


# --- run ------------------------------------------------------------------

BACKGROUND_RUN_STATUSES: Final[tuple[str, ...]] = (
    "queued",
    "running",
    "succeeded",
    "retryable_failed",
    "terminal_failed",
    "skipped",
    "cancelled",
    "superseded",
)

BackgroundRunStatus = Literal[
    "queued",
    "running",
    "succeeded",
    "retryable_failed",
    "terminal_failed",
    "skipped",
    "cancelled",
    "superseded",
]


class BackgroundRun(BaseModel):
    """One execution attempt. ``errorMessage`` is sanitized — only a stable
    class/code, never a raw third-party exception."""

    model_config = ConfigDict(extra="forbid")

    id: str = Field(min_length=1)
    uid: str = Field(min_length=1)
    scheduleId: str = Field(min_length=1)  # noqa: N815
    kind: BackgroundWorkflowKind
    status: BackgroundRunStatus
    idempotencyKey: str = Field(min_length=1)  # noqa: N815
    scheduledFor: str  # noqa: N815
    inputWindowStart: str  # noqa: N815
    inputWindowEnd: str  # noqa: N815
    startedAt: str | None = None  # noqa: N815
    finishedAt: str | None = None  # noqa: N815
    attempt: int = Field(ge=0)
    leaseExpiresAt: str | None = None  # noqa: N815
    outputRef: str | None = Field(default=None, min_length=1)  # noqa: N815
    errorCode: str | None = Field(default=None, min_length=1)  # noqa: N815
    errorMessage: str | None = None  # noqa: N815
    model: str | None = Field(default=None, min_length=1)
    tokenCostEstimate: float | None = Field(default=None, ge=0)  # noqa: N815
    createdAt: str  # noqa: N815

    @field_validator("scheduledFor", "inputWindowStart", "inputWindowEnd", "createdAt")
    @classmethod
    def required_timestamps(cls, v: str) -> str:
        return _validate_iso8601(v, "timestamp")

    @field_validator("startedAt", "finishedAt", "leaseExpiresAt")
    @classmethod
    def optional_timestamps(cls, v: str | None) -> str | None:
        return v if v is None else _validate_iso8601(v, "timestamp")


# --- proposed action ------------------------------------------------------

PROPOSED_ACTION_TYPES: Final[tuple[str, ...]] = (
    "archive_message",
    "create_task",
    "create_calendar_event",
)

ProposedActionType = Literal["archive_message", "create_task", "create_calendar_event"]

PROPOSED_ACTION_STATUSES: Final[tuple[str, ...]] = (
    "proposed",
    "approved",
    "rejected",
    "executed",
    "failed",
)

ProposedActionStatus = Literal["proposed", "approved", "rejected", "executed", "failed"]


class ProposedActionResult(BaseModel):
    model_config = ConfigDict(extra="forbid")

    ok: bool
    ref: str | None = Field(default=None, min_length=1)
    errorCode: str | None = Field(default=None, min_length=1)  # noqa: N815


class BackgroundProposedAction(BaseModel):
    """A write the run proposes but never performs — it needs foreground
    confirmation before routing to the existing Workspace write tools."""

    model_config = ConfigDict(extra="forbid")

    id: str = Field(min_length=1)
    uid: str = Field(min_length=1)
    runId: str = Field(min_length=1)  # noqa: N815
    notificationId: str | None = Field(default=None, min_length=1)  # noqa: N815
    type: ProposedActionType
    status: ProposedActionStatus
    sourceMessageIds: list[str]  # noqa: N815
    summary: str = Field(min_length=1)
    params: dict[str, Any] | None = None
    result: ProposedActionResult | None = None
    createdAt: str  # noqa: N815

    @field_validator("sourceMessageIds")
    @classmethod
    def source_ids_non_empty(cls, v: list[str]) -> list[str]:
        for mid in v:
            if not mid:
                raise ValueError("sourceMessageIds must be non-empty strings")
        return v

    @field_validator("createdAt")
    @classmethod
    def created_at_iso(cls, v: str) -> str:
        return _validate_iso8601(v, "createdAt")


# --- notification / digest ------------------------------------------------

NOTIFICATION_STATUSES: Final[tuple[str, ...]] = (
    "unread",
    "read",
    "dismissed",
    "acted_on",
)

NotificationStatus = Literal["unread", "read", "dismissed", "acted_on"]

NotificationItemBucket = Literal["noise", "actions", "events", "info"]


class BackgroundNotificationItem(BaseModel):
    """One row in a digest — stable Workspace IDs + a short snippet only,
    never the full body/address."""

    model_config = ConfigDict(extra="forbid")

    messageId: str = Field(min_length=1)  # noqa: N815
    threadId: str | None = Field(default=None, min_length=1)  # noqa: N815
    bucket: NotificationItemBucket
    subject: str = Field(min_length=1)
    snippet: str = Field(min_length=1)


class BackgroundNotification(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str = Field(min_length=1)
    uid: str = Field(min_length=1)
    runId: str = Field(min_length=1)  # noqa: N815
    kind: BackgroundWorkflowKind
    status: NotificationStatus
    title: str = Field(min_length=1)
    summary: str = Field(min_length=1)
    items: list[BackgroundNotificationItem]
    proposedActions: list[str]  # noqa: N815
    createdAt: str  # noqa: N815
    expiresAt: str | None = None  # noqa: N815

    @field_validator("proposedActions")
    @classmethod
    def action_ids_non_empty(cls, v: list[str]) -> list[str]:
        for aid in v:
            if not aid:
                raise ValueError("proposedActions must be non-empty strings")
        return v

    @field_validator("createdAt")
    @classmethod
    def created_at_iso(cls, v: str) -> str:
        return _validate_iso8601(v, "createdAt")

    @field_validator("expiresAt")
    @classmethod
    def expires_at_iso(cls, v: str | None) -> str | None:
        return v if v is None else _validate_iso8601(v, "expiresAt")


# --- task-id sanitization -------------------------------------------------

_INVALID_TASK_ID_CHARS = re.compile(r"[^A-Za-z0-9_-]")


def sanitize_task_id(value: str) -> str:
    """Mirror of ``sanitizeTaskId`` in background.ts. Cloud Tasks task IDs
    allow only letters, numbers, hyphens, and underscores (ADR §4); replace
    every other character with ``_``. Pure character replacement — callers
    hash long/sensitive identifiers before composing the id."""
    return _INVALID_TASK_ID_CHARS.sub("_", value)

File: packages/shared-types/src/background.ts
import { z } from 'zod';

/**
 * Contracts for the background / scheduled agent-work subsystem (ADR 0001).
 *
 * These shapes cross the web↔agent boundary: the Python agent owns the
 * dispatcher + executor and writes `backgroundRuns` / `backgroundNotifications`
 * / proposed-action records; the web app reads schedules + digests. Both sides
 * import the same contract — the Pydantic mirror lives in
 * `apps/agent_py/src/lifecoach_agent/contracts/background.py` and the parity
 * test (`tests/unit/test_background_contracts.py`) keeps them in lock-step.
 *
 * Timestamps are ISO-8601 strings on the wire (`z.string().datetime()`), even
 * though Firestore stores them as native Timestamps server-side — the adapter
 * converts at the storage boundary.
 *
 * OAuth tokens, raw email bodies, addresses, and full snippets must NEVER
 * appear in any of these records (ADR §Security). The schemas only carry IDs,
 * counts, status, sanitized error classes, and short client-safe text.
 */

// --- workflow kinds -------------------------------------------------------

/**
 * Workflow kinds the subsystem can schedule. The first release ships only
 * `email_triage_daily`; `email_urgent_scan` is reserved for the next
 * workflow and kept here so the enum (and the Firestore index) is stable.
 */
export const BACKGROUND_WORKFLOW_KINDS = ['email_triage_daily', 'email_urgent_scan'] as const;

export type BackgroundWorkflowKind = (typeof BACKGROUND_WORKFLOW_KINDS)[number];

// --- schedule -------------------------------------------------------------

/** How aggressive a per-action policy is allowed to be (ADR §1). The first
 * release treats every third-party write as `after_confirmation` at most;
 * `auto_if_rule_matches` is reserved for a future ADR. */
export const PERMITTED_ACTION_MODES = [
  'never',
  'after_confirmation',
  'auto_if_rule_matches',
] as const;

export type PermittedActionMode = (typeof PERMITTED_ACTION_MODES)[number];

/** Bounded Gmail lookback windows for a triage run. */
export const LOOKBACK_WINDOWS = ['12h', '1d', '3d'] as const;

export type LookbackWindow = (typeof LOOKBACK_WINDOWS)[number];

/** Last summary status surfaced on the schedule doc for the settings UI. */
export const SCHEDULE_LAST_STATUSES = ['ok', 'skipped', 'failed'] as const;

export type ScheduleLastStatus = (typeof SCHEDULE_LAST_STATUSES)[number];

/** Daily cadence policy. `localTime` is `HH:MM` 24h in the schedule's IANA
 * `timezone`; `weekdays` (0=Sunday … 6=Saturday, JS `Date.getDay()`) narrows
 * to specific days — omit for every day. An empty array is rejected: it would
 * mean "no days" (the schedule never legitimately fires), so it's an invalid
 * config rather than a silently-degraded one. */
export const ScheduleCadenceSchema = z
  .object({
    type: z.literal('daily'),
    localTime: z.string().regex(/^([01]\d|2[0-3]):[0-5]\d$/, 'localTime must be HH:MM 24h'),
    weekdays: z
      .array(z.number().int().min(0).max(6))
      .min(1, 'weekdays must be omitted or non-empty')
      .optional(),
  })
  .strict();

export type ScheduleCadence = z.infer<typeof ScheduleCadenceSchema>;

export const PermittedActionsSchema = z
  .object({
    archiveNoise: z.enum(PERMITTED_ACTION_MODES),
    createTasks: z.enum(PERMITTED_ACTION_MODES),
    createCalendarEvents: z.enum(PERMITTED_ACTION_MODES),
  })
  .strict();

export type PermittedActions = z.infer<typeof PermittedActionsSchema>;

export const NotifyPreferencesSchema = z
  .object({
    inApp: z.boolean(),
    email: z.boolean(),
    chatSummaryOnNextOpen: z.boolean(),
  })
  .strict();

export type NotifyPreferences = z.infer<typeof NotifyPreferencesSchema>;

/** True if `tz` is a resolvable IANA timezone. Rejects "PST" / "not-a-zone"
 * which would mis-compute local daily runs + DST transitions. */
function isValidTimeZone(tz: string): boolean {
  try {
    new Intl.DateTimeFormat('en-US', { timeZone: tz });
    return true;
  } catch {
    return false;
  }
}

/** A user-configured automation. Application state, not model state. */
export const BackgroundScheduleSchema = z
  .object({
    id: z.string().min(1),
    uid: z.string().min(1),
    kind: z.enum(BACKGROUND_WORKFLOW_KINDS),
    enabled: z.boolean(),
    timezone: z.string().min(1).refine(isValidTimeZone, 'timezone must be a valid IANA zone'),
    cadence: ScheduleCadenceSchema,
    lookbackWindow: z.enum(LOOKBACK_WINDOWS),
    consentVersion: z.string().min(1),
    permittedActions: PermittedActionsSchema,
    notify: NotifyPreferencesSchema,
    nextRunAt: z.string().datetime(),
    lastRunAt: z.string().datetime().optional(),
    lastStatus: z.enum(SCHEDULE_LAST_STATUSES).optional(),
    createdAt: z.string().datetime(),
    updatedAt: z.string().datetime(),
  })
  .strict();

export type BackgroundSchedule = z.infer<typeof BackgroundScheduleSchema>;

// --- run ------------------------------------------------------------------

/** Lifecycle of one background run (ADR §6). The app owns terminal state —
 * Cloud Tasks is never relied on as a dead-letter queue. */
export const BACKGROUND_RUN_STATUSES = [
  'queued',
  'running',
  'succeeded',
  'retryable_failed',
  'terminal_failed',
  'skipped',
  'cancelled',
  'superseded',
] as const;

export type BackgroundRunStatus = (typeof BACKGROUND_RUN_STATUSES)[number];

/** A single execution attempt record. `errorMessage` is sanitized — only a
 * stable class/code, never a raw third-party exception (ADR §Error
 * sanitization). */
export const BackgroundRunSchema = z
  .object({
    id: z.string().min(1),
    uid: z.string().min(1),
    scheduleId: z.string().min(1),
    kind: z.enum(BACKGROUND_WORKFLOW_KINDS),
    status: z.enum(BACKGROUND_RUN_STATUSES),
    // `{scheduleId}:{kind}:{scheduledFor}` — persisted for auditability even
    // when the public Cloud Tasks task ID is hashed/shortened (ADR §4).
    idempotencyKey: z.string().min(1),
    scheduledFor: z.string().datetime(),
    inputWindowStart: z.string().datetime(),
    inputWindowEnd: z.string().datetime(),
    startedAt: z.string().datetime().optional(),
    finishedAt: z.string().datetime().optional(),
    attempt: z.number().int().min(0),
    leaseExpiresAt: z.string().datetime().optional(),
    outputRef: z.string().min(1).optional(),
    errorCode: z.string().min(1).optional(),
    errorMessage: z.string().optional(),
    model: z.string().min(1).optional(),
    tokenCostEstimate: z.number().nonnegative().optional(),
    createdAt: z.string().datetime(),
  })
  .strict();

export type BackgroundRun = z.infer<typeof BackgroundRunSchema>;

// --- proposed action ------------------------------------------------------

/** A write the run proposes but never performs — it needs foreground
 * confirmation before routing to the existing Workspace write tools. */
export const PROPOSED_ACTION_TYPES = [
  'archive_message',
  'create_task',
  'create_calendar_event',
] as const;

export type ProposedActionType = (typeof PROPOSED_ACTION_TYPES)[number];

export const PROPOSED_ACTION_STATUSES = [
  'proposed',
  'approved',
  'rejected',
  'executed',
  'failed',
] as const;

export type ProposedActionStatus = (typeof PROPOSED_ACTION_STATUSES)[number];

/** Result of executing a proposed action once it's approved in foreground.
 * `ref` is the created/changed Workspace resource id; `errorCode` is a stable
 * sanitized class, never a raw API message. */
export const ProposedActionResultSchema = z
  .object({
    ok: z.boolean(),
    ref: z.string().min(1).optional(),
    errorCode: z.string().min(1).optional(),
  })
  .strict();

export type ProposedActionResult = z.infer<typeof ProposedActionResultSchema>;

export const BackgroundProposedActionSchema = z
  .object({
    id: z.string().min(1),
    uid: z.string().min(1),
    runId: z.string().min(1),
    notificationId: z.string().min(1).optional(),
    type: z.enum(PROPOSED_ACTION_TYPES),
    status: z.enum(PROPOSED_ACTION_STATUSES),
    // Stable Workspace message IDs the action derives from; never raw content.
    // At least one — an auditable action must tie back to a concrete message.
    sourceMessageIds: z.array(z.string().min(1)).min(1),
    // Short client-safe description ("Archive 4 newsletters from last week").
    summary: z.string().min(1),
    // Action-specific arguments (e.g. task title, event start). Schema-free
    // by design — the foreground approval flow maps these onto the existing
    // Workspace write tools, which own their own arg validation.
    params: z.record(z.string(), z.unknown()).optional(),
    result: ProposedActionResultSchema.optional(),
    createdAt: z.string().datetime(),
  })
  .strict();

export type BackgroundProposedAction = z.infer<typeof BackgroundProposedActionSchema>;

// --- notification / digest ------------------------------------------------

export const NOTIFICATION_STATUSES = ['unread', 'read', 'dismissed', 'acted_on'] as const;

export type NotificationStatus = (typeof NOTIFICATION_STATUSES)[number];

/** One row in a digest. Carries stable Workspace IDs + a short snippet only —
 * enough for the user to recognise the message without re-opening Gmail, and
 * never the full body/address (ADR §Security). */
export const BackgroundNotificationItemSchema = z
  .object({
    messageId: z.string().min(1),
    threadId: z.string().min(1).optional(),
    bucket: z.enum(['noise', 'actions', 'events', 'info']),
    subject: z.string().min(1),
    snippet: z.string().min(1),
  })
  .strict();

export type BackgroundNotificationItem = z.infer<typeof BackgroundNotificationItemSchema>;

export const BackgroundNotificationSchema = z
  .object({
    id: z.string().min(1),
    uid: z.string().min(1),
    runId: z.string().min(1),
    kind: z.enum(BACKGROUND_WORKFLOW_KINDS),
    status: z.enum(NOTIFICATION_STATUSES),
    title: z.string().min(1),
    summary: z.string().min(1),
    items: z.array(BackgroundNotificationItemSchema),
    // IDs of `BackgroundProposedAction` records requiring foreground confirm.
    proposedActions: z.array(z.string().min(1)),
    createdAt: z.string().datetime(),
    expiresAt: z.string().datetime().optional(),
  })
  .strict();

export type BackgroundNotification = z.infer<typeof BackgroundNotificationSchema>;

// --- task-id sanitization -------------------------------------------------

/**
 * Cloud Tasks task IDs allow only letters, numbers, hyphens, and underscores
 * (ADR §4). Replace every other character (colons in ISO timestamps, dots,
 * slashes, `@` in raw uids, …) with `_` so a deterministic id built from
 * schedule/kind/timestamp is always a valid task name.
 *
 * This is intentionally pure character replacement — it does NOT hash or
 * shorten; callers hash long/sensitive identifiers (truncated SHA-256) before
 * composing the id, then pass the composed string here as a final guard.
 */
export function sanitizeTaskId(input: string): string {
  return input.replace(/[^A-Za-z0-9_-]/g, '_');
}
